dxr/plugins/python/indexers.py

```diff
@@ -249,44 +257,43 @@ def analyze_tokens(self):
             if tok_type == token.NAME:
                 # AST nodes for classes and functions point to the position of
                 # their 'def' and 'class' tokens. To get the position of their
-                # names, we look for 'def' and 'class' tokens and store the
-                # position of the token immediately following them.
-                if node_start and node_type == 'definition':
-                    node_start_table[node_start[0]] = (start, end)
-                    node_type, node_start = None, None
-                    continue
-
-                if tok_name in ('def', 'class'):
-                    node_type, node_start = 'definition', (utf8_start, start)
-                    continue
-
-                # Record all name nodes in the token table.  Currently unused,
-                # but will be needed for recording variable references.
-                node_start_table[utf8_start] = (start, end)
-                node_type, node_start = 'name', (utf8_start, start)
+                # names, we start the queue for the current parenthesis level
+                # at the byte offset for the keyword token, but only start
+                # pushing character offsets once we're past the keyword.
+                paren_stack.setdefault(paren_level, (utf8_start, []))
+
+                if tok_name not in ('def', 'class'):
+                    paren_stack[paren_level][1].append((start, end))
+
+                continue
 
             elif tok_type == token.OP:
-                # In order to properly capture the start and end of function
-                # calls, we need to keep track of the parens.  Put the
-                # starting positions on a stack (here implemented with a dict
-                # so that it can be sparse), but only if the previous node was
-                # a name.
-                if tok_name == '(':
-                    if node_type == 'name':
-                        paren_stack[paren_level] = node_start
+                # Container delimiters (parens, brackets, and braces) start a
```

Comment:
Really, I don't see anything functionally wrong with the code. The tests are very encouraging; it looks like this now gets a bunch more stuff right! I'm just having trouble understanding it. In particular, I could use some explanatory comments about the purpose of paren_stack. Are nodes inside parens coming out of the tokenizer with positions relative to those parens or something? Once I can get that through my thick Friday-afternoon skull, we can merge this. Thanks so much for the brain-tremor-inducing-but-apparently-correct patch, Jeff! :-)